### DISCUSSION TITLE
Mobile Otp login

### DISCUSSION THREAD

**Discussion by bisverse** [saleor/saleor#9890] [Q&A]
are there any way to sign up using mobile number (no email) and later verify user with otp. I looked into oidc documentation but it requires email as response.

**Comment by manojLondhe** [comment]
This is challenging, but **doable**.

We are currently working on creating an Android app (on top of GQL APIs + custom APIs we wrote externally) that needs mobile  + OTP based login

**We ended up implementing this as below:**
- We are using [OIDC plugin by Saleor](https://github.com/saleor/saleor/tree/main/saleor/plugins/openid_connect) 
- We are using [Keycloak](https://www.keycloak.org/) as OPENID server
- We are using [Saleor as a client for OPENID server](https://docs.saleor.io/docs/3.x/developer/available-plugins/openid-connect#saleor-as-a-client-for-authorization-server)
- We added custom code for keycloak to support mobile + OTP based login [this is the challenging part]

So,
- we first register new user using mobile number + other fields in Keycloak using Keycloak REST API
- then we use mutations described [here](https://docs.saleor.io/docs/3.x/developer/available-plugins/openid-connect#saleor-as-a-client-for-authorization-server) to generate token in the APP
- when you use OIDC plugin mentioned above, plugin creates new user in Saleor with basic details from token received from OPENID server (Keycloak)

There is one problem though:
- Since Saleor needs email ID for all accounts we end up putting dummy email ids based on mobile for all users in keycloak eg: 9876543210@mydomain.local

Except dummy email IDs all is working fine for us.

**Reply by bisverse** [reply]
Thanks a lot, I think this is a good solution.

**Reply by ghost** [reply]
Hi Manoj, any way I could contact you to get a better idea of the solution you've implemented? I am looking to use phone authentication for an application but saleor doesn't have this feature built in..

**Reply by peteror** [reply]
we implemented a much simpler thing, running in production for almost a year. I think Keycloak is quite a heavy solution just for this.  (we have an Android + iOS app)

- we have a middleware app that generates, sends and validates the OTP (the OIDC plugin was not available when we started). Our mobile app calls the API of this middleware for login.
- the middleware  logs in or registers customers in Saleor (where we store the phone number as phonenumber@whatever.com just like Manoj - works well ) using the GraphQL APIs
- we have a database where we store hashed (random, string) passwords that we log customers in with using this middleware.
